datafusion/functions-array/src/lib.rs

```diff
@@ -139,13 +326,16 @@ pub fn register_all(registry: &mut dyn FunctionRegistry) -> Result<()> {
         replace::array_replace_all_udf(),
         replace::array_replace_udf(),
     ];
+
+    // TODO: Remove. Define in array_functions directly
     functions.into_iter().try_for_each(|udf| {
```

Comment:
I think you could avoid the duplication here by looping over the elements of `array_functions` and instantiating them all

For example (untested)

```rust
array_functions()
  .values()
  .try_for_each(|udf_factory| {
    let udf = (udf_factory)();
    ..
  });
```

Reply:
We are register functions here, so instantiation should we avoided. Only when we call `get_udf`, is time we instantiate the function.

Reply:
Maybe it is worth clarifying if the goal is to when we are trying to avoi instantiating `ScalarUDF` 

It seems like it could be either:
1. When we have a physical plan already (aka as in Comet) and need to look up a function by name (but not iterate through the entire list)
2. Any time a `SessionContext` is created (aka make FunctionFactory handle deferred registrations too).

I can see the usecase for 1, I am not sure about for 2 🤔

Reply:
@viirya Do you have suggestions on which usecase is your preferred one?